eth-rpc/client/clientdispatch.nim

```diff
@@ -77,37 +82,17 @@ proc connect*(self: RpcClient, address: string, port: Port) {.async.} =
   self.port = port
   asyncCheck processData(self)
 
-proc makeTemplate(name: string, params: NimNode, body: NimNode, starred: bool): NimNode =
-  # set up template AST
-  result = newNimNode(nnkTemplateDef)
-  if starred: result.add postFix(ident(name), "*")
-  else: result.add ident(name)
-  result.add newEmptyNode(), newEmptyNode(), params, newEmptyNode(), newEmptyNode(), body
-
-proc appendFormalParam(formalParams: NimNode, identName, typeName: string) =
-  # set up formal params AST
-  formalParams.expectKind(nnkFormalParams)
-  if formalParams.len == 0: formalParams.add newEmptyNode()
-  var identDef = newIdentDefs(ident(identName), ident(typeName))
-  formalParams.add identDef
-
 macro generateCalls: untyped =
   ## Generate templates for client calls so that:
   ##   client.call("web3_clientVersion", params)
   ## can be written as:
   ##   client.web3_clientVersion(params)
   result = newStmtList()
   for callName in ETHEREUM_RPC_CALLS:
-    # REVIEW: `macros.quote` would have worked well here to make the code easier to understand/maintain
-    var
-      params = newNimNode(nnkFormalParams)
-      call = newCall(newDotExpr(ident("client"), ident("call")), newStrLitNode(callName), ident("params"))
-      body = newStmtList().add call
-      templ = makeTemplate(callName, params, body, true)
-    params.add newNimNode(nnkBracketExpr).add(ident("Future"), ident("Response"))
-    params.appendFormalParam("client", "RpcClient")
-    params.appendFormalParam("params", "JsonNode")
-    result.add templ
+    let nameLit = ident(callName)
+    result.add(quote do:
+      template `nameLit`*(client: RpcClient, params: JsonNode): Future[Response] = client.call(`callName`, params)  # TODO: Back to template
```

Comment:
I've updated `ethprocs` to use type signatures, so although the root RPC is accepting a `JsonNode`, the `on` macro processes this into Nim types.

All the RPCs accept Json arrays for parameters, and the spec says the RPC should accept a Json array for it's parameters. As such we should only need one `JsonNode` as a parameter, but the `on` macro allows clearer type specifications.
For example: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getbalance

I can upload the updated `ethprocs`, but unfortunately I have a bug to fix in `servertypes.nim` somewhere, so it's complaining about not being able to find `result` - in the past this has actually been because it couldn't process a parameter type, and I think currently this is caused by something returning a non type when the macro is expecting an expression.

Reply:
@zah see the updated `ethprocs` I just uploaded.

Reply:
I see what you mean about the client procs though. Would be nice to have them specified as the server ones are.

Reply:
Clients can now use native nim calls for ethprocs, which are generated from the signatures in `ethsigs.nim`.
These automatically marshal data from Nim to json types, perform the call, then marshal any return types from json back into native Nim types.